File: src/testLint/testUtils.ts
/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { SpawnSyncOptions, spawnSync } from 'child_process'

export function runCmd(args: string[], options?: SpawnSyncOptions & { throws?: boolean }) {
    const result = spawnSync(args[0], args.slice(1), options)

    const throws = options?.throws ?? true
    if (throws && result.status !== 0) {
        throw new Error(`
-----
Error running: $ ${args.join(' ')}

status: ${result.status}
error: ${result.error?.toString()}
stdout: ${result.stdout?.toString()}
stderr: ${result.stderr?.toString()}
-----`)
    }
    return result
}

File: plugins/core/core/tst/software/aws/toolkits/core/utils/test/TestUtils.kt
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.core.utils.test

import java.time.Duration
import java.time.Instant
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger
import kotlin.random.Random

fun aString(length: Int = Random.nextInt(5, 30)): String = UUID.randomUUID().toString().substring(length)

fun retryableAssert(
    timeout: Duration? = null,
    maxAttempts: Int? = null,
    interval: Duration = Duration.ofMillis(100),
    assertion: () -> Unit
) {
    val calculatedTimeout = timeout ?: maxAttempts?.let { Duration.ofMillis(it * (interval.toMillis() * 2)) } ?: Duration.ofSeconds(1)
    val expiry = Instant.now().plus(calculatedTimeout)
    val attempts = AtomicInteger(0)
    while (true) {
        try {
            assertion()
            return
        } catch (e: AssertionError) { // deliberately narrowed to an AssertionError - this is intended to be used in a test assertion
            when {
                Instant.now().isAfter(expiry) -> throw e
                maxAttempts != null && attempts.incrementAndGet() >= maxAttempts -> throw e
                else -> Thread.sleep(interval.toMillis())
            }
        }
    }
}
